rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -635,6 +652,61 @@ private predicate typeEquality(AstNode n1, TypePath prefix1, AstNode n2, TypePat
   prefix2.isEmpty()
 }
 
+/**
+ * Holds if `child` is a child of `parent`, and the Rust compiler applies [least
+ * upper bound (LUB) coercion](1) to infer the type of `parent` from the type of
+ * `child`.
+ *
+ * In this case, we want type information to only flow from `child` to `parent`,
+ * to avoid (a) either having to model LUB coercions, or (b) risk combinatorial
+ * explosion in inferred types.
```

Comment:
So if we're not computing LUC coercions, what happens (for the type of `a`) in a case such as:
```
let a = if b {
  let x = &v;
  x
} else {
  let y = &mut v;
  y
}
```

Reply:
`a` will get the type of both `x` and `y` (as was the case before this PR).

Reply:
OK, I guess that's something we could do better in future but is orthogonal to this PR.